syntax = "proto3";

import "types.proto";

package flame;

option go_package = "github.com/flame-sh/flame/sdk/go/rpc";

message ApplicationContext {
    string name = 1;
    Shim shim = 2;
    optional string image = 3;
    optional string command = 4;
    optional string working_directory = 5;
}

message SessionContext {
    string session_id = 1;
    ApplicationContext application = 2;
    optional bytes common_data = 3;
}

message TaskContext {
    string task_id = 1;
    string session_id = 2;
    optional bytes input = 4;
}

message WatchEventResponse {
    EventOwner owner = 1;
    Event event = 2;
}

service Instance {
    rpc OnSessionEnter(SessionContext) returns (Result) {}
    rpc OnTaskInvoke(TaskContext) returns (TaskResult) {}
    rpc OnSessionLeave(EmptyRequest) returns (Result) {}

    rpc WatchEvent(EmptyRequest) returns (stream WatchEventResponse) {}
}
